// https://github.com/grpc/grpc/blob/v1.6.x/examples/protos/route_guide.proto
// with some modification

syntax = "proto3";

package routeguide;

service RouteGuide
{
  // A simple RPC.
  // Obtains the feature at a given position.
  rpc GetFeature(Point) returns (Feature) {}

  // A server-to-client streaming RPC.
  // Obtains the Features available within the given Rectangle.  
  rpc ListFeatures(Rectangle) returns (stream Feature) {}

  // A client-to-server streaming RPC.
  // Accepts a stream of Points on a route being traversed, returning a
  // RouteSummary when traversal is completed.
  rpc RecordRoute(stream Point) returns (RouteSummary) {}

  // A Bidirectional streaming RPC.
  // Accepts a stream of RouteNotes sent while a route is being traversed,
  // while receiving other RouteNotes (e.g. from other users).
  rpc RouteChat(stream RouteNote) returns (stream RouteNote) {}
}

message Point {
  int32 latitude = 1;
  int32 longitude = 2;
}

// A latitude-longitude rectangle, represented as two diagonally opposite
message Rectangle {
  Point lo = 1;
  Point hi = 2;
}

// A feature names something at a given point.
// If a feature could not be named, the name is empty.
message Feature {
  string name = 1;
  Point location = 2;
}

// A RouteNote is a message sent while at a given point.
message RouteNote {
  Point location = 1;
  string message = 2;
}

// A RouteSummary is received in response to a RecordRoute rpc.
message RouteSummary {
  int32 point_count = 1;
  int32 distance = 2;
  int32 elapsed_time = 3;
  string note = 4;
}